WordPress/gutenberg
# Fix keyboard navigation on the Image block toolbar

Fixes #24766

This PR addresses the case where replacing parts of the block toolbar "on the fly" breaks keyboard navigation when the new buttons aren't using `ToolbarItem` or `ToolbarButton` components, like in the Image block toolbar.

It should've shown a deprecation warning, but [this change](https://github.com/WordPress/gutenberg/pull/23425#issuecomment-660339641) had introduced a regression where the block toolbar stopped tracking new elements inside it. The changes in the `navigable-toolbar` component are related to this.

It doesn't fix the focus loss issue, which is a separate problem that will be fixed in a follow-up PR. Currently, the user would have to press <kbd>Shift+Tab</kbd> to go back to the toolbar after clicking on the crop button.

### Screenshots

![Using the crop function in the Image block toolbar with the keyboard](https://user-images.githubusercontent.com/3068563/93477179-0db46180-f8d1-11ea-8853-4445553e734c.gif)

### How to test

- Create an image block.
- Activate the crop button.
- Check if the toolbar works with the keyboard as expected.

---

### Site editor

Currently, the block toolbar doesn't support text fields. In the Site Editor, the Template parts toolbar adds a text field asynchronously. Now it fallbacks to the legacy toolbar. It shows a deprecation warning that has been ignored in the tests for now.

<img src="https://user-images.githubusercontent.com/3068563/93479326-8ff15580-f8d2-11ea-95a6-f1b99f65edfd.png" alt="Template parts block toolbar with a text field inside" height="110" />
